NOTIF-7 Notifications table layout & edit forms

josejulio opened the pull request:
![Peek 2020-09-22 08-16](https://user-images.githubusercontent.com/3845764/93887207-0794fb00-fcac-11ea-8c21-9671f53d0266.gif)

![Screenshot_2020-09-30_12-51-06](https://user-images.githubusercontent.com/3845764/94721563-af996c80-031b-11eb-8104-ddc1a7ff58e6.png)
![Screenshot_2020-09-30_12-51-20](https://user-images.githubusercontent.com/3845764/94721571-b4f6b700-031b-11eb-9891-b5a65eb79afa.png)

bastilian commented:
The problem of endpoints moving to the bottom of the list is addressed with reducer improvements in https://github.com/RedHatInsights/notifications-frontend/pull/21/files#diff-98635d199a7b94b3612ddcd02d62519aR19, where endpoints are updated inplace in an object and will stay in their current row in the table as well and not change position.

adamruzicka commented:
Closing in favor of #21

josejulio committed:
NOTIF-7 Starting the notifications table layout

josejulio committed:
Styling and grouping

josejulio committed:
Preparation for expandend content, possible bug or something wrong is there.

josejulio committed:
Adding more filters, start the Default behavior

josejulio committed:
Adding "default behavior" on top of the table and the export button

josejulio committed:
Allow no-actions on notifications

josejulio committed:
Add edit notification/default actions modal

josejulio committed:
Fix lint errors

josejulio committed:
WIP - Edit form
 - Need to split several components into their own files and put some definitions

josejulio committed:
Adding the integration action as recipient in the modal form

josejulio committed:
Adding the recipient and action with integration on table / default behavior

josejulio committed:
Split EditableActionTable into several files

josejulio committed:
Select are appended to document.body
 Allows the Select to be used inside modals without being clipped

AllenBW commented on src/app/App.scss:
```diff
@@ -6,3 +6,7 @@
 }
+
+.pf-c-page__main#root {
+  --pf-c-page__main-section--BackgroundColor: #FFF;
```
might want to use the pf global var here `var(--pf-global--Color--light-100)` (same value)

AllenBW commented on src/components/Notifications/ActionComponent.tsx:
```diff
@@ -0,0 +1,76 @@
+import * as React from 'react';
+import { BellIcon, EnvelopeIcon, SlackHashIcon } from '@patternfly/react-icons';
```
might want to grab these individually from the /dist/ems (or js) folder means we don't have to bring in whole lib

AllenBW replied:
something like `import BellIcon from '@patternfly/react-icons/dist/ems/icons/bell-icon';`

josejulio replied:
I though this was done already (grabbing individual code) but looks like it doesn't... I wonder if there are other ways to avoid the full-syntax per import.

AllenBW commented on src/components/Notifications/ActionComponent.tsx:
```diff
@@ -0,0 +1,76 @@
+}
+
+const marginLeftClassName = style({
+    marginLeft: Spacer.SM
```
pf constants are always preferred to local ones... alllssooooooo m a yuge fan of separating style from other code (some scss somewhere) but not sure if either of those changes are easily manageable with line and 23

josejulio replied:
I prefer to have the code connected when dealing with jsx/tsx. Is easier for me to quickly see where this class is applied and remove stale elements. The linter will usually let me know if this goes unused or if I'm using a wrong value in there.

AllenBW commented on src/components/Notifications/ActionComponent.tsx:
```diff
@@ -0,0 +1,76 @@
+        );
+    }
+
+    if (props.action === undefined) {
```
```suggestion
    if (!props.action) {
```

AllenBW commented on src/components/Notifications/ActionComponent.tsx:
```diff
@@ -0,0 +1,76 @@
+    if (props.action === undefined) {
+        return (
+            <ActionComponentWrapper { ...props }>
+                <span className={ grayFontClassName }>
```
might want to use the pf text component here... https://patternfly-react.surge.sh/components/text

josejulio replied:
Could you elaborate about using that component? I suppose I'm missing something from it.

AllenBW commented on src/components/Notifications/DefaultBehavior.tsx:
```diff
@@ -0,0 +1,71 @@
+    onEdit: () => void;
+}
+
+const contentClassName = style({
```
same comments about styling separation n using pf constants as before

AllenBW commented on src/components/Notifications/DefaultBehavior.tsx:
```diff
@@ -0,0 +1,71 @@
+import { getOuiaProps } from '../../utils/getOuiaProps';
+import { style } from 'typestyle';
+import { ActionComponent } from './ActionComponent';
+import { Button, ButtonVariant, Flex, FlexItem } from '@patternfly/react-core';
```
these can also be imported piecemeal from ah `/dist/ems(umd/js)`

AllenBW replied:
Flex, ie layouts from `@patternfly/react-core/dist/ems/layouts/`

AllenBW commented on src/components/Notifications/Form.tsx:
```diff
@@ -0,0 +1,178 @@
+    Notification,
+    IntegrationRef
+} from '../../types/Notification';
+import { Button, ButtonVariant } from '@patternfly/react-core';
```
same dist comments as above

AllenBW commented on src/components/Notifications/Form.tsx:
```diff
@@ -0,0 +1,178 @@
+    getIntegrations: (type: IntegrationType, search: string) => Array<IntegrationRef>;
+}
+
+const alignLeftClassName = style({
```
same style things as before, constants n seperated

AllenBW commented on src/components/Notifications/DefaultBehavior.tsx:
```diff
@@ -0,0 +1,71 @@
+
+export const DefaultBehavior: React.FunctionComponent<DefaultBehaviorProps> = (props) => {
+    return (
+        <div { ...getOuiaProps('Notifications/DefaultBehavior', props) } className={ contentClassName } >
```
yah can kill the 
```
{
    return (
```
from above n just return this

AllenBW commented on src/components/Notifications/SaveModal.tsx:
```diff
@@ -0,0 +1,66 @@
+}
+
+const InternalNotificationSaveModal: React.FunctionComponent<InternalProps> = (props) => {
+    const title = props.type === 'default' ? 'Edit default notification actions' : 'Edit notification actions';
```
```suggestion
    const title =  `Edit${props.type === 'default' && ' default'} notification actions;
```

AllenBW commented on src/components/Notifications/Table.tsx:
```diff
@@ -0,0 +1,346 @@
+});
+
+const noBorderBottom = style({
+    borderBottom: 'none !important'
```
never fun to rely on !important, importing ah .scss directly into this component would likely eliminate the need for this

(or maybe not because yer build pipeline is different? 🤔 )

josejulio replied:
The issue I encountered with this syntax is that PF was added *after* these kind of declarations, thus PF definition won here.
I suppose that some way to counter this is to check if we can bump the priority of typestyle.

AllenBW commented on src/components/Notifications/Table.tsx:
```diff
@@ -0,0 +1,346 @@
+                            pfBorderBottomClassName
+                        ),
+                        colSpan: parent === undefined ? 1 : 1 // Todo: Change this to `? 1 : 2` once the PF bug is fixed
+                        // https://github.com/patternfly/patternfly-react/issues/4858
```
❤️

AllenBW commented on src/components/Notifications/Table.tsx:
```diff
@@ -0,0 +1,346 @@
+
+        if (parent !== undefined) {
+            rows[rows.length - 1].parent = parent;
+            // rows[rows.length - 1].useAllCellInExpandedContent = true;
```
comment we can kill?

AllenBW commented on src/components/Notifications/Table.tsx:
```diff
@@ -0,0 +1,346 @@
+
+            if (parent !== undefined) {
+                rows[rows.length - 1].parent = parent;
+                // rows[rows.length - 1].useAllCellInExpandedContent = true;
```
comment we can kill?

AllenBW commented on src/components/Notifications/Table.tsx:
```diff
@@ -0,0 +1,346 @@
+};
+
+const toTableRowsGroupedByApplication = (applicationGroups: Array<NotificationRowGroupedByApplication>, onEdit: OnEditNotification): Array<IRow> => {
+    return applicationGroups.reduce((rows, applicationGroup) => {
```
can kill the `{ return `n  save a line, return the `applicationGroups.` ...

AllenBW commented on src/pages/Notifications/List/Page.tsx:
```diff
@@ -1,27 +1,288 @@
+                integration: {
+                    type: IntegrationType.WEBHOOK,
+                    id: 'integration-00001',
+                    name: 'Send stuff over there'
```
😆

AllenBW commented on package.json:
```diff
@@ -4,22 +4,23 @@
     "csstips": "^1.2.0",
     "csx": "^10.0.1",
     "date-fns": "^2.9.0",
+    "emotion": "^10.0.27",
```
professional preference, lockin down all deps to exact versions, less surprises 🤷 🏆 :trollface: 🤗

josejulio replied:
Actually this shouldn't even be here. Had to do some imports when testing with patternfly/react-core. I'll just remove this.

josejulio replied:
> lockin down all deps to exact versions

Ah, I never got to finish locking all the dependencies, that's something I have to do.

AllenBW reviewed (commented):
HIIIIII this is exciting, gotta lotta question marks about how this all works together that m workin on piecing together... but here are some high level thoughts...

ah few feedbacks got a lil repetitive so i stopped doin em, but they apply everywhere, using distribution versions of deps, styling separate from and consumed by components (rather than baked in) and then then use of pf variables for all styling things (when possible)...

🥇 💯 🍍

josejulio committed:
Apply suggestions from code review

Co-authored-by: Allen Wight <allen.b.wight@gmail.com>

josejulio committed:
Some review comments

josejulio committed:
- Add missing close template string `
- Fixed a typo

josejulio commented:
Merging this to be able to have something to see on the notifications page, some review recommendations need changes on the whole app or common lib.
